src/tools/cgset.c

```diff
@@ -77,11 +77,139 @@ static int cgroup_set_cgroup_values(struct cgroup *src_cgrp, const char * const
 		goto err;
 	}
 
+	return cgrp;
+
+err:
+	cgroup_free(&cgrp);
+	return NULL;
+}
+
+static int cgroup_set_cgroup_values(struct cgroup *src_cgrp, const char * const new_cgrp)
+{
+	struct cgroup *cgrp;
+	int ret = ECGFAIL;
+
+	cgrp = create_and_copy_cgroup(src_cgrp, new_cgrp);
+	if (!cgrp)
+		return ret;
+
 	/* modify cgroup based on values of the new one */
 	ret = cgroup_modify_cgroup(cgrp);
 	if (ret)
 		err("%s: cgroup modify error: %s\n", program_name, cgroup_strerror(ret));
 
+	cgroup_free(&cgrp);
+	return ret;
+}
+
+static int _cgroup_set_cgroup_values_r(struct cgroup *src_cgrp, const char * const new_cgrp)
+{
+	struct cgroup_file_info info;
+	int prefix_len;
+	void *handle;
+	int lvl, ret;
+
+
+	ret = cgroup_walk_tree_begin(src_cgrp->controller[0]->name, new_cgrp, 0, &handle, &info,
+			&lvl);
+	if (ret != 0)
+		return ret;
+
+	prefix_len = strlen(info.full_path) - strlen(new_cgrp) - 1;
+	ret = cgroup_set_cgroup_values(src_cgrp, &info.full_path[prefix_len]);
+	if (ret)
+		goto err;
+
+	while ((ret = cgroup_walk_tree_next(0, &handle, &info, lvl)) == 0) {
+		if (info.type == CGROUP_FILE_TYPE_DIR) {
+
+			ret = cgroup_set_cgroup_values(src_cgrp, &info.full_path[prefix_len]);
+			if (ret)
+				goto err;
+		}
+	}
+
+err:
+	if (ret == ECGEOF)
+		ret = 0; /* we successfully walked the tree */
+
+	cgroup_walk_tree_end(&handle);
+	return ret;
+}
+
+static int cgroup_copy_controller_idx(struct cgroup *dst_cgrp, struct cgroup *src_cgrp, int idx)
+{
+	struct cgroup_controller *src_cgc = src_cgrp->controller[idx];
+	struct cgroup_controller *dst_cgc = NULL;
+	int ret, i;
+
+	dst_cgc = cgroup_add_controller(dst_cgrp, src_cgc->name);
+	if (!dst_cgc) {
+		err("%s: failed to add controller %s to %s\n", program_name, src_cgc->name,
+							       dst_cgrp->name);
+		return ECGFAIL;
+	}
+
+	for (i = 0; i < src_cgc->index; i++) {
+		ret = cgroup_add_value_string(dst_cgc, src_cgc->values[i]->name,
+					      src_cgc->values[i]->value);
+		if (ret) {
+			err("%s: Failed to add value %s to cgroup %s controller %s\n",
+			    program_name, src_cgc->values[i]->name, dst_cgrp->name, src_cgc->name);
+			return ret;
+		}
+	}
+
+	return 0;
+}
+
+static int cgroup_set_cgroup_values_r(struct cgroup *src_cgrp, const char * const new_cgrp)
+{
+	struct cgroup *cgrp;
+	int i, ret = ECGFAIL;
+
+	if (is_cgroup_mode_unified()) {
```

Comment:
What about hybrid mode?

Reply:
Yeah, we need to take in account the `hybrid()` mode too, will fix it in the next version.